wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/DefaultFactoryParameterResolver.java

```diff
@@ -56,10 +64,15 @@ public boolean accept(Map<String, String> factoryParameters) {
   public FactoryDto createFactory(@NotNull final Map<String, String> factoryParameters)
       throws BadRequestException, ServerException {
     // create factory from the following devfile location
+    String devfileLocation = factoryParameters.get(URL_PARAMETER_NAME);
+
+    FileContentProvider contentProvider =
+        devfileLocation == null
```

Comment:
how it can be `null`  here ?

Reply:
Actually not sure. What I do know is that there is a null check for this in `createFactoryFromDevfile`. Maybe we could remove that null check and move it here? There don't seem to be any other usages of `createFactoryFromDevfile` so I'm hoping we should be safe.

Reply:
The problem is deeper here. If you remove that check, than which provider you will always  pass to the next mehod?
`URLFileContentProvider`?  Than `FetchNotSupportedProvider` wil be never used & message never shown. So in fact it is depends of URL user provided in factory. Some of them is applicable for changing the last filename part, some are not.  As a variant, i can propose `createFactoryFromDevfile` to accept an array of content providers, and iterate over them,  so find one which gives content or throws `IOexception`. And if no one succeeds, the last one (FetchNotSupported) will throw an DevfileExc.

Reply:
I would just not call that method at all if I wasn't able to provide it with a devfile location...

Reply:
It is not a question where to reject if it is no devfile URL. It is question how to disctincts betwee those two content providers?

Reply:
So the `createFactoryFromDevfile` needs the the devfile location so that it can fetch its YAML. It makes no sense to call it with null for the location. So we can assume that the devfileLocation is always non-null and ask the caller to not even attempt calling it with null location.

As for the file content provider, it is up to the caller to provide an impl to `createFactoryFromDevfile`. The caller knows it can only call the method if the devfile location is known and it also knows how to resolve the files referenced in the devfile. In case of the default behavior, we just use the URI.resolve() against the devfile location, in case of github we basically use URL translation to the raw content URLs of github.

Notice that nowhere in the above, I needed the `FetchNotSupportedProvider`. Now that I understand the logic a little bit more, I think it is actually not needed at all. We do support some sort of file content resolution in all cases.

Reply:
I went ahead and did the changes I had in mind that basically go inline with your original finding that `devfileLocation` can never be null here. If you don't agree with that, please comment further on the newer version of the code.